x-pack/elastic-agent/pkg/agent/install/uninstall.go

```diff
@@ -177,3 +193,147 @@ func programsFromConfig(cfg *config.Config) ([]program.Program, error) {
 
 	return pp, nil
 }
+
+func applyDynamics(ctx context.Context, log *logger.Logger, cfg *config.Config) (*config.Config, error) {
+	cfgMap, err := cfg.ToMapStr()
+	ast, err := transpiler.NewAST(cfgMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// apply dynamic inputs
+	inputs, ok := transpiler.Lookup(ast, "inputs")
+	if ok {
+		varsArray := make([]*transpiler.Vars, 0, 0)
+		var wg sync.WaitGroup
+		wg.Add(1)
+		varsCallback := func(vv []*transpiler.Vars) {
+			varsArray = vv
+			wg.Done()
+		}
+
+		ctrl, err := composable.New(log, cfg)
+		if err != nil {
+			return nil, err
+		}
+		ctrl.Run(ctx, varsCallback)
+		wg.Wait()
+
+		renderedInputs, err := renderInputs(inputs, varsArray)
+		if err != nil {
+			return nil, err
+		}
+		err = transpiler.Insert(ast, renderedInputs, "inputs")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// apply caps
+	caps, err := capabilities.Load(paths.AgentCapabilitiesPath(), log, status.NewController(log))
+	if err != nil {
+		return nil, err
+	}
+
+	astIface, err := caps.Apply(ast)
+	if err != nil {
+		return nil, err
+	}
+
+	newAst, ok := astIface.(*transpiler.AST)
+	if ok {
+		ast = newAst
+	}
+
+	finalConfig, err := newAst.Map()
+	return config.NewConfigFrom(finalConfig)
+}
+
+// Dynamic inputs section
+// TODO(michal): move to shared code during refactoring
+func renderInputs(inputs transpiler.Node, varsArray []*transpiler.Vars) (transpiler.Node, error) {
```

Comment:
I know you marked this as TODO. But I don't think we should land this duplicated. This code path is identical to the emitter piece.

With input templates this will need to be updated again. The duplication will most likely result in this code path being forgot or missed.

Reply:
let me think what we can do here

Reply:
Being this is going to go into 7.12. We can leave it, but lets quickly clean this up in a follow up.

Reply:
moved to transpiler, works good